bob-apps/bin/bobc.rs

```diff
@@ -1,71 +1,376 @@
+                Some(KeyPattern::Single(val)) => {
+                    let key = get_key_value(val, app_args.keysize);
+                    exist(vec![key], addr).await;
+                }
+                Some(KeyPattern::Range(range)) => {
+                    // FIXME: empty response when range is descending (14-1)
+                    let keysize = app_args.keysize;
+                    let keys = range.map(|v| get_key_value(v, keysize)).collect();
+                    exist(keys, addr).await;
+                }
+                Some(KeyPattern::Multiple(vec)) => {
+                    let keysize = app_args.keysize;
+                    let keys = vec.into_iter().map(|v| get_key_value(v, keysize)).collect();
+                    exist(keys, addr).await;
+                }
+                None => unreachable!()
             }
-            _ => {}
         }
+        _ => unreachable!(),
     }
 }
 
-async fn put(key: Vec<u8>, size: usize, addr: Uri) {
-    let mut client = BobApiClient::connect(addr).await.unwrap();
+async fn prepare_put_from_pattern(re_path: &Regex, dir: &PathBuf, key_size: usize) -> Vec<KeyName> {
+    let mut dir_iter = read_dir(dir).await.unwrap();
 
-    let timestamp = SystemTime::now()
-        .duration_since(UNIX_EPOCH)
-        .expect("msg: &str")
-        .as_secs();
-    let meta = BlobMeta { timestamp };
-    let blob = Blob {
-        data: vec![1; size],
-        meta: Some(meta),
-    };
-    let message = PutRequest {
-        key: Some(BlobKey { key }),
-        data: Some(blob),
-        options: None,
-    };
-    let put_req = Request::new(message);
+    let mut keys_names = Vec::new();
+    while let Some(entry) = dir_iter.next_entry().await.unwrap() {
+        let file = entry.path();
+        if !file.is_dir() && re_path.is_match(file.to_str().unwrap()) {
+            let name = file.to_str().unwrap().to_owned();
+            let key = &re_path.captures(&name).unwrap()[1];
+            let key = get_key_value(key.parse().unwrap(), key_size);
+            keys_names.push(KeyName { key, name });
+        }
+    }
+    info!("{:?}", keys_names);
+    keys_names
+}
 
-    let res = client.put(put_req).await;
-    info!("{:#?}", res);
+fn prepare_get(filenames: Vec<(u64, String)>, key_size: usize) -> Vec<KeyName> {
+    let mut keys_names = Vec::with_capacity(filenames.len());
+    for (i, name) in filenames {
+        keys_names.push(KeyName {
+            key: get_key_value(i, key_size),
+            name,
+        })
+    }
+    keys_names
 }
 
-async fn get(key: Vec<u8>, addr: Uri) {
+async fn put(key: Vec<u8>, filename: &str, addr: Uri) {
+    let file = File::open(filename).await;
+
+    match file {
+        Err(e) => {
+            error!("{:?}", e);
+        }
+        Ok(mut f) => {
+            let mut client = BobApiClient::connect(addr).await.unwrap();
+
+            let timestamp = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .expect("msg: &str")
```

Comment:
Expect msg seems misleading

Reply:
This was not my code but I thought there was a reason for that. Seems not. I can just unwrap this value: [649b3b6](https://github.com/qoollo/bob/pull/562/commits/649b3b62130876d19f2c6b45c504dc01f2c199fe)